pkg/controller/component/service_descriptor_utils.go

```diff
@@ -32,132 +33,239 @@ import (
-		return nil, nil
+func referencedServiceVars(ctx context.Context, cli client.Reader, namespace string,
+	serviceRef appsv1alpha1.ServiceRef) (*appsv1alpha1.CredentialVar, *appsv1alpha1.CredentialVar, error) {
+	var (
+		clusterRef     = serviceRef.ClusterRef
+		endpoint, port *appsv1alpha1.CredentialVar
+	)
+
+	if clusterRef.Service == nil {
+		return nil, nil, nil
 	}
 
-	if len(compDef.Spec.ServiceRefDeclarations) == 0 {
-		return nil, nil
+	serviceKey := types.NamespacedName{Namespace: namespace}
+	if serviceRef.Namespace != "" {
+		serviceKey.Namespace = serviceRef.Namespace
+	}
+	if len(clusterRef.Service.Component) == 0 {
+		serviceKey.Name = constant.GenerateClusterServiceName(clusterRef.Cluster, clusterRef.Service.Service)
+	} else {
+		generate := constant.GenerateComponentServiceName
+		if clusterRef.Service.Service == "headless" {
+			generate = constant.GenerateComponentHeadlessServiceName
+		}
+		serviceKey.Name = generate(clusterRef.Cluster, clusterRef.Service.Component, clusterRef.Service.Service)
 	}
 
-	serviceReferences := make(map[string]*appsv1alpha1.ServiceDescriptor, len(compDef.Spec.ServiceRefDeclarations))
-	for _, serviceRefDecl := range compDef.Spec.ServiceRefDeclarations {
-		for _, serviceRef := range comp.Spec.ServiceRefs {
-			if serviceRef.Name != serviceRefDecl.Name {
-				continue
-			}
-			targetNamespace := namespace
-			if serviceRef.Namespace != "" {
-				targetNamespace = serviceRef.Namespace
-			}
-			// if service reference is another KubeBlocks Cluster, then it is necessary to generate a service connection credential from the cluster connection credential secret
-			if serviceRef.Cluster != "" {
-				if err := handleClusterTypeServiceRef(reqCtx, cli, targetNamespace, clusterName, serviceRef, serviceRefDecl, serviceReferences); err != nil {
-					return nil, err
-				}
-				// serviceRef.Cluster takes precedence, and if serviceRef.Cluster is set, serviceRef.ServiceDescriptor will be ignored
-				break
-			}
+	service := &corev1.Service{}
+	if err := cli.Get(ctx, serviceKey, service); err != nil {
+		return nil, nil, err
+	}
 
-			if serviceRef.ServiceDescriptor != "" {
-				if err := handleServiceDescriptorTypeServiceRef(reqCtx, cli, targetNamespace, serviceRef, serviceRefDecl, serviceReferences); err != nil {
-					return nil, err
-				}
+	resolvePort := func() *int32 {
+		if len(service.Spec.Ports) == 0 {
+			return nil
+		}
+		if len(clusterRef.Service.Port) == 0 {
+			return &service.Spec.Ports[0].Port
+		}
+		for i, p := range service.Spec.Ports {
+			if p.Name == clusterRef.Service.Port {
+				return &service.Spec.Ports[i].Port
 			}
 		}
-		// _, exist := serviceReferences[serviceRefDecl.Name]
-		// if !exist {
-		//	 return nil, fmt.Errorf("componentDef %s's serviceRefDeclaration %s has not been defined, please check if there is corresponding service definition and binding in Cluster.spec.componentSpecs[*].serviceRefs", clusterCompDef.Name, serviceRefDecl.Name)
-		// }
+		return nil
```

Comment:
If the port specified in clusterRef.Service.Port does not exist, should it raise an error?

Reply:
fixed